src/box/lua/call.c

```diff
@@ -93,74 +93,80 @@ get_call_serializer(void)
 }
 
 /**
- * A helper to find a Lua function by name and put it
- * on top of the stack.
+ * A helper to resolve a Lua function by full name, for example like:
+ * foo.bar['biz']["baz"][3].object:function
+ * Puts the function on top of the stack, followed by an object (if present).
+ * Returns number of items pushed (1 or 2) or -1 in case of error (diag is set).
  */
 static int
 box_lua_find(lua_State *L, const char *name, const char *name_end)
 {
-	int index = LUA_GLOBALSINDEX;
-	int objstack = 0, top = lua_gettop(L);
-	const char *start = name, *end;
-
-	while ((end = (const char *) memchr(start, '.', name_end - start))) {
-		lua_checkstack(L, 3);
-		lua_pushlstring(L, start, end - start);
-		lua_gettable(L, index);
-		if (! lua_istable(L, -1)) {
-			diag_set(ClientError, ER_NO_SUCH_PROC,
-				 name_end - name, name);
-			return -1;
-		}
-		start = end + 1; /* next piece of a.b.c */
-		index = lua_gettop(L); /* top of the stack */
-	}
+	lua_checkstack(L, 3); /* No more than 3 entries are needed. */
+	int top = lua_gettop(L);
 
-	/* box.something:method */
-	if ((end = (const char *) memchr(start, ':', name_end - start))) {
-		lua_checkstack(L, 3);
-		lua_pushlstring(L, start, end - start);
-		lua_gettable(L, index);
-		if (! (lua_istable(L, -1) ||
-			lua_islightuserdata(L, -1) || lua_isuserdata(L, -1) )) {
-				diag_set(ClientError, ER_NO_SUCH_PROC,
-					  name_end - name, name);
-				return -1;
+	/* Take the first token. */
+	const char *start = name;
+	while (start != name_end && strchr(".:[", *start) == NULL)
+		start++;
+	lua_pushlstring(L, name, start - name);
+	lua_gettable(L, LUA_GLOBALSINDEX);
+
+	/* Take the rest tokens. */
+	while (start != name_end) {
+		if (!lua_istable(L, -1) &&
+		    !lua_islightuserdata(L, -1) && !lua_isuserdata(L, -1))
+			goto no_such_proc;
+
+		char delim = *start++; /* skip delimiter. */
+		if (delim == '.') {
+			/* Look for the next token. */
+			const char *end = start;
+			while (end != name_end && strchr(".:[", *end) == NULL)
+				end++;
+			lua_pushlstring(L, start, end - start);
+			start = end;
+		} else if (delim == ':') {
+			lua_pushlstring(L, start, name_end - start);
+			lua_gettable(L, -2); /* get function from object. */
+			lua_insert(L, -2); /* swap function and object. */
+			break;
+		} else {
+			assert(delim == '[');
+			const char *end = memchr(start, ']', name_end - start);
+			if (end == NULL ||
+			    (end != name_end && strchr(".:[", end[1]) == NULL))
+				goto no_such_proc;
+
+			if (end - start >= 2 && start[0] == end[-1] &&
+			    (start[0] == '"' || start[0] == '\'')) {
+				/* Quoted string, just extract it. */
+				lua_pushlstring(L, start + 1, end - start - 2);
+			} else {
+				/* Must be a number, convert from string. */
+				lua_getfield(L, LUA_GLOBALSINDEX, "tonumber");
+				lua_pushlstring(L, start, end - start);
+				lua_call(L, 1, 1);
```

Comment:
I'd rather use `lua_isnumber` and `lua_tonumber` here.

Reply:
Thanks, applied the following patch:
```diff
diff --git a/src/box/lua/call.c b/src/box/lua/call.c
index e0e4cbd34f..05780ab93d 100644
--- a/src/box/lua/call.c
+++ b/src/box/lua/call.c
@@ -101,7 +101,7 @@ get_call_serializer(void)
 static int
 box_lua_find(lua_State *L, const char *name, const char *name_end)
 {
-       lua_checkstack(L, 3); /* No more than 3 entries are needed. */
+       lua_checkstack(L, 2); /* No more than 2 entries are needed. */
        int top = lua_gettop(L);
 
        /* Take the first token. */
@@ -143,9 +143,13 @@ box_lua_find(lua_State *L, const char *name, const char *name_end)
                                lua_pushlstring(L, start + 1, end - start - 2);
                        } else {
                                /* Must be a number, convert from string. */
-                               lua_getfield(L, LUA_GLOBALSINDEX, "tonumber");
                                lua_pushlstring(L, start, end - start);
-                               lua_call(L, 1, 1);
+                               int success;
+                               lua_Number num = lua_tonumberx(L, -1, &success);
+                               if (!success)
+                                       goto no_such_proc;
+                               lua_pop(L, 1);
+                               lua_pushnumber(L, num);
                        }
                        start = end + 1; /* skip closing bracket. */
                }
```